weather_dl/download_pipeline/manifest.py

```diff
@@ -242,22 +444,138 @@ def _update(self, download_status: DownloadStatus) -> None:
+            bigquery.SchemaField('download_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the download state ends."),
+            bigquery.SchemaField('upload_start_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the upload state starts."),
+            bigquery.SchemaField('upload_end_time', 'TIMESTAMP', mode='NULLABLE',
+                                 description="A UTC datetime when the upload state ends."),
+        ]
+        table = bigquery.Table(self.location, schema=TABLE_SCHEMA)
+        with bigquery.Client() as client:
+            client.create_table(table, exists_ok=True)
+
+    def _read(self, location: str) -> DownloadStatus:
+        """Reads the JSON data from a manifest."""
+        with bigquery.Client() as client:
+            query = f"SELECT * FROM {self.location} WHERE location = {location!r}"
+            query_job = client.query(query)
+            result = query_job.result().to_dataframe().to_dict('records')
+            row = {n: to_json_serializable_type(v) for n, v in result[0].items()}
+            return DownloadStatus.from_dict(row)
+
+    def _update(self, download_status: DownloadStatus) -> None:
+        """Writes the JSON data to a manifest."""
+        with bigquery.Client() as client:
+            status = DownloadStatus.to_dict(download_status)
+            table = client.get_table(self.location)
+            columns = [field.name for field in table.schema]
+            update_dml = []
+            insert_dml = []
+            for col in columns:
+                if status[col] is None:
+                    update_dml.append(f"{col} = null")
+                    insert_dml.append("null")
+                elif col == 'selection':
+                    update_dml.append(f"{col} = JSON'{status[col]}'")
+                    insert_dml.append(f"JSON'{status[col]}'")
+                elif col == 'size' or col == 'error':
+                    update_dml.append(f"{col} = {status[col]}")
+                    insert_dml.append(status[col])
+                else:
+                    update_dml.append(f"{col} = '{status[col]}'")
+                    insert_dml.append(f"'{status[col]}'")
+
+            # Build the merge statement as a string
+            merge_statement = f"""
+                MERGE {self.location} T
+                USING (
+                SELECT
+                    '{status['location']}' as location
+                ) S
+                ON T.location = S.location
+                WHEN MATCHED THEN
+                UPDATE SET
+                    {', '.join(str(v) for v in update_dml)}
+                WHEN NOT MATCHED THEN
+                INSERT
+                    ({", ".join(columns)})
+                VALUES
+                    ({', '.join(str(v) for v in insert_dml)})
+            """
```

Comment:
Does BQ include some sort of protection against SQL injection? I think this system of formatting strings is unsafe.

Reply:
It is necessary to use a library to sanitize user inputs.

Reply:
Thanks Alex for pointing this.

I have made the necessary changes for this -- by using parameterized queries, we ensure that the user input is properly sanitized and prevent SQL injection attacks, because the user input is never directly embedded in the SQL query string.